### DISCUSSION TITLE
How to handle a library that will only build dynamically, in a static triplet?

### DISCUSSION THREAD

**Discussion by jdx-john** [microsoft/vcpkg#20076] [Q&A] [answered]
I've got my new port (https://github.com/alanxz/SimpleAmqpClient) to build on windows against the default `x86-windows` triplet, but our build process uses the StaticMD (static library linkage, dynamic CRT) and the upstream library simply doesn't support static linkage.

I _think_ I've seen before that some existing ports are DLL-only, but they still build (as DLL, with a warning) when you do a static build - is this possible? I can live with one library producing a DLL _if_ it's just one. When I build that library manually using CMake, it seems happy to statically link its dependencies (basically boost and librabbitmqc) into the DLL but how can I configure this in a portfile?

And for that matter, how can I explicitly set a C++ preprocessor value? I tried to use `add_definitions(-DMYDEF)`

But I got error 

> CMake Error: Command add_definitions() is not scriptable

So I thought maybe I can do:
```
vcpkg_configure_cmake(
  SOURCE_PATH ${SOURCE_PATH}
  PREFER_NINJA
  OPTIONS
	-DMYDEF=1
)
```
But this just (I think) sets a _CMake_ variable, not a preprocessor variable. I think it must be easy to do?

**Comment by Osyotr** [comment, marked as answer]
In that case you should put `vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)` at the top of your portfile. See [docs](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_check_linkage.md).

Example:
https://github.com/microsoft/vcpkg/blob/master/ports/qt5-webengine/portfile.cmake

**Reply by jdx-john** [reply]
That seems like what I was after - will check tomorrow and close (I hope).
